services/worker-manager/src/providers/aws.js

```diff
@@ -0,0 +1,342 @@
+    });
+
+    //todo: do something to enable logging to stackdriver
+  }
+
+  async provision({workerPool}) {
+    const {workerPoolId} = workerPool;
+
+    if (!workerPool.providerData[this.providerId] || workerPool.providerData[this.providerId].running === undefined) {
+      await workerPool.modify(wt => {
+        wt.providerData[this.providerId] = wt.providerData[this.providerId] || {};
+        wt.providerData[this.providerId].running = wt.providerData[this.providerId].running || 0; // why zero?
+      });
+    }
+
+    const toSpawn = await this.estimator.simple({
+      workerPoolId,
+      ...workerPool.config,
+      running: workerPool.providerData[this.providerId].running,
+    });
+
+    let spawned;
+
+    try {
+      spawned = await this.ec2.runInstances({
+        IamInstanceProfile: {
+          Arn: this.instanceProfile.Arn,
+          Name: this.instanceProfile.InstanceProfileName,
+        },
+        MaxCount: toSpawn,
+        MinCount: toSpawn,
+        ImageId: workerPool.config.imageId,
+        InstanceType: workerPool.config.instanceType,
+        Placement: {
+          AvailabilityZone: this.region,
+        },
+        TagSpecifications: {
+          ResourceType: 'instance',
+          Tags: [{
+            Key: 'Provider',
+            Value: `wm-${this.providerId}`,
+          }, {
+            Key: 'Owner',
+            Value: workerPool.owner,
+          }, {
+            Key: 'Project',
+            Value: this.project,
+          }],
+        },
+      });
+    } catch (e) {
+      await workerPool.reportError({
+        kind: 'creation-error',
+        title: 'Instance Creation Error',
+        description: e.message, // TODO: Make sure we clear exposing this with security folks
+        notify: this.notify,
+        WorkerPoolError: this.WorkerPoolError,
+      });
+    }
+
+    Promise.all(spawned.Instances.map(i => {
+      return this.Worker.create({
+        workerPoolId,
+        providerId: this.providerId,
+        workerGroup: this.providerId,
+        workerId: i.InstanceId,
+        created: new Date(),
+        expires: taskcluster.fromNow('1 week'),
+        state: this.Worker.states.REQUESTED,
+        providerData: { // what do we want to remember, anyways? 🤔
+          project: this.project,
+        },
+      });
+    }));
+  }
+
+  async deprovision({workerPool}) {
+    // should this be implemented in Provider? Looks like it's going to be the same for all
+    await workerPool.modify(wt => {
+      wt.previousProviderIds = wt.previousProviderIds.filter(p => p !== this.providerId);
+      delete wt.providerData[this.providerId];
+    });
+    // why don't we remove workers from db or terminate instances?
+  }
+
+  async registerWorker({worker, workerPool, workerIdentityProof}) {
+    // check worker's identity -
+    // The way AWS works, workers will be getting temporary creds for accessing AWS APIs
+    // automatically. There's no token back-and-forth
```

Comment:
The worker will need to provide something to worker-manager to prove that it really is this worker -- otherwise anyone can call `workerManager.registerWorker` and get Taskcluster credentials for free.  When John was thinking about this he was looking at using [instance identity documents](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instance-identity-documents.html) for this purpose.

Reply:
oh thank you! I'll look into that, I was so confused about this. I had thought this method just helped getting google creds to call the google APIs

Reply:
oh, I actually was looking at that documentation page on Thursday/Friday. So my understanding is, the worker doesn't need to even perform `curl http://169.254.169.254/latest/dynamic/instance-identity/document` - it's done automatically. Because I create a role with policies in `setup`, and that is attached to each instance through `InstanceProfile` https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_use_switch-role-ec2.html

You are mentioning the Taskcluster credentials though. What is the workflow there? I find myself not knowing which credentials will the workers need and how they usually get them

Reply:
> You are mentioning the Taskcluster credentials though. What is the workflow there? I find myself not knowing which credentials will the workers need and how they usually get them

Details are [in the docs](https://github.com/taskcluster/taskcluster/blob/master/services/worker-manager/providers.md#registering) but `registerWorker` is an API method that is called without any Taskcluster credentials, and returns Taskcluster credentials.  So it has to include *something* to prove that the caller is who they claim to be.  I'm not sure how that will work for AWS.

Reply:
[More docs here](https://docs.taskcluster.net/docs/reference/core/worker-manager/worker-interaction) :)

Reply:
Thank you!